src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs

```diff
@@ -1112,26 +1112,32 @@ private SymbolicRegexNode<TSet> PruneLowerPriorityThanNullability(SymbolicRegexB
 
                 case SymbolicRegexNodeKind.Concat:
                     Debug.Assert(_left is not null && _right is not null);
-                    //in a concatenation (X|Y)Z priority is given to XZ when X is nullable
-                    //observe that (X|Y)Z is equivalent to (XZ|YZ) and the branch YZ must be ignored
-                    //when X is nullable (observe that XZ is nullable because this=(X|Y)Z is nullable)
-                    //if X is not nullable then XZ is maintaned as is, and YZ is pruned
-                    //e.g. (a{0,5}?|abc|b+)c* reduces to c*
-                    //---
-                    //in a concatenation XZ where X is not an alternation, both X and Z are pruned
-                    //e.g. a{0,5}?b{0,5}? reduces to ()
-                    prunedNode = _left._kind == SymbolicRegexNodeKind.Alternate ?
-                        (_left._left!.IsNullableFor(context) ?
+                    prunedNode = _left._kind switch
+                    {
+                        // If the left side is a concatenation, then we bring it to right associative form to expose
+                        // the first element of the concatenation for the cases below.
+                        SymbolicRegexNodeKind.Concat => CreateConcat(builder, _left._left!, CreateConcat(builder, _left._right!, _right))
+                            .PruneLowerPriorityThanNullability(builder, context),
+                        // In a concatenation (X|Y)Z priority is given to XZ when X is nullable.
+                        // Observe that (X|Y)Z is equivalent to (XZ|YZ) and the branch YZ must be ignored
+                        // when X is nullable (observe that XZ is nullable because this=(X|Y)Z is nullable).
```

Comment:
> observe that XZ is nullable because this=(X|Y)Z is nullable

I didn't understand this part.  Why is `(X|Y)Z` nullable?  Is it just that we're only here in this code if `this` node is known to be nullable, and in this example, this is `(X|Y)Z`?

Reply:
In general, a concatenation `WZ` is nullable **iff** both `W` and `Z` are nullable, so its holds as a direct implication then that both `W=(X|Y)` and `Z` are nullable, therefor if `X` is nullable the so is `XZ`.